docs/man/tcdecode.1
.TH tcdecode 1 "23th August 2003" "tcdecode(1)"
.SH NAME
tcdecode \- read multimedia streams from medium, decode to raw format and print to standard output
.SH SYNOPSIS
.TP
.B tcdecode
.B -i
.I name
[
.B -x
.I codec
] [
.B -t
.I package
] [
.B -g
.I WxH
] [
.B -s
.I c,f,r
] [
.B -C
.I s-e
] [
.B -y
.I codec
] [
.B -Y
] [
.B -Q
.I mode
] [
.B -A
.I flag
] [
.B -d
.I verbosity
] [
.B -v
]
.SH COPYRIGHT
\fBtcdecode\fP is Copyright (C) by Thomas Oestreich.
.SH DESCRIPTION
.B tcdecode
is part of and usually called by \fBtranscode\fP.
.br
However, it can also be used independently.
.br
\fBtcdecode\fP reads source (from stdin if not explicitely defined) and
prints on the standard output.
.SH OPTIONS
.IP "\fB-i \fIname\fP"
Specify input source.  If ommited, \fIstdin\fP is assumed.
\fBtcdecode\fP reads streams from file or from stdin.

.IP "\fB-g \fIWxH\fP"
video frame width and height [720x576]

.IP "\fB-C \fIs-e\fP"
decode and print out only this frame interval (video) or bytes (audio) [all].

.IP "\fB-x \fIcodec\fP"
source codec, if not detected of reading from stdin. This option also specifies video and audio context of the source for decoding. Currently, supported parameter for \fIcodec\fP are

.RS 8
.TS
lI c l.
mp3	MPEG audio	(lame)
ac3	AC3 audio	(liba52)
dv	Digital Video DV	(libdv)
mpeg2	MPEG video	(libmpeg2)
yv12	Y'CbCr YUV
mov	Quicktime	(libquicktime)
.TE
.RE

.IP "\fB-t \fIpackage\fP"
Currently only argument supported is "lavc". Use the codec by -x from this package. Available (-x) codecs for \fIlavc\fP are:
mpeg2, divx3, divx, divx4, mp42, mjpg, rv10, svq1, svq3, vp3, 4xm, wmv1, wmv2, hfyu, indeo3, h263p, h263i


.IP "\fB-y \fIcodec\fP"
target codec. \fB tcdecode \fP only decodes to raw format suitable for transcode. Valid option parameter for \fIcodec\fP are

.RS 8
.TS
lI r.
rgb	RGB 24bit
yv12	Y'CbCr YUV
yuy2	4:2:2 YUV
pcm	SIGNED LE 16bit PCM
.TE
.RE
.IP "\fB-s\fP \fIc,f,r\fP"
audio gain for AC3 downmixing [1.0,1.0,1.0]

.IP "\fB-Y\fP"
decoded Digital Video (raw) YUV frame is in YUY2 (packet) format using libdv. Downsample frame to YV12. PAL users should compile libdv with --with-pal-yuv=YV12 to avoid this option [off]

.IP "\fB-A\fP \fIflag\fP"
audio flag for AC3/A52 decoder [none]. This flag determines the down-mixing
configuration. Valid choices for \fIflag\fP are determined by the following
bits set:

.RS 8
.TS
l r.
drc off	1
demux	2
dolby off	4
.TE
.RE

Add those numbers to turn multiple options on.

.IP "\fB-Q\fP \fImode\fP"
decoding quality. Certain codecs, e.g., DV with libdv allow for changing
this parameter if speed is prefered over quality.
Valid parameter are 1=fastest-5=best [5].

.IP "\fB-d\fP \fIlevel\fP"
With this option you can specify a bitmask to enable different levels
of verbosity (if supported).  You can combine several levels by adding the
corresponding values:

.RS 8
.TS
l r.
QUIET	0
INFO	1
DEBUG	2
STATS	4
WATCH	8
FLIST	16
VIDCORE	32
SYNC	64
COUNTER	128
PRIVATE	256
.TE
.RE

.IP "\fB-v\fP"
Print version information and exit.

.IP "\fB-X\fP \fItype\fP"
enforce experimental IA32/AMD64 acceleration using a comma-separated list of types [autodetect].
\fItype\fR may any one of
.RS 8
.TS
lI c l.
C	No acceleration
asm	plain x86 assembly
mmx	MMX instruction set
3dnow	3DNow! instruction set
sse	SSE instruction set
sse2	SSE2 instruction set
.TE
.RE
.RE

.SH NOTES
\fBtcdecode\fP is a front end for decoding various sources and is used in \fBtranscode(1)\fP's import modules.
.SH EXAMPLES
The command
.nf
\	tcdecode -i foo.mp3 -x mp3
.fi
decodes an MPEG audio stream and prints raw PCM data to stdout.

The command
.nf
\	tcextract -i foo.avi -x mp3 | tcdecode -x mp3
.fi
extracts an MPEG audio track from the AVI-file foo.avi and pipes the frames into the decoder

.PP
.SH AUTHORS
.B tcdecode
was written by Thomas Oestreich
.br
<ostreich@theorie.physik.uni-goettingen.de> with contributions from
many others.  See AUTHORS for details.
.SH SEE ALSO
.BR avifix (1),
.BR avisync (1),
.BR avimerge (1),
.BR avisplit (1),
.BR tcprobe (1),
.BR tcscan (1),
.BR tccat (1),
.BR tcdemux (1),
.BR tcextract (1),
.BR tcdecode (1),
.BR transcode (1)
.br


ChangeLog
Key:
[+] feature added
[!] bugfix of an existing feature
[*] change/improvement (not bugfix) of an existing feature
[-] feature dropped

===========================================================================
transcode-1.2.0
---------------
[+] Brand new export layer with new-style, separate encoders and muxers.
[+] New encoders: theora, vorbis, faac, x264.
[+] New multiplexors: ogg, libavformat, yuv4mpeg2.
[+] New filter: barrel.
[+] ShoutCast support using ogg multiplexor.
[!] Fix crash when using transformations (-Z, -j).
[!] Fix compilation error with x264 version 78.
[*] Improved internal buffering system.
[*] Added basic A/V synchronization module.
[!] Fix misuse of configuration files which leads to crash.
[*] Allow to handle fallback directories for configuration files.
[*] Improved debug messages handling and internal debug API.
[-] The import_imlist is deprecated. Use --multi_input -x im instead.
[*] Big YAIT update by Allan Snider.
[!] Make the -W option work again like 1.0.x branch.
[*] Switch from ImageMagick to GraphicsMagick.
[*] Allow to build without FFmpeg.
[!] Remove outdated libav* information from the FAQ.
[+] Added support for CubicKeys4 and Sinc8 zoom algorithms.
[+] Added module registry to automatically select a module from format
[+] Added `tccfgshow' to inspect the transcode internal capabilities.
[-] Removed the unmantained PVM code.
[+] Added a special-crafted import module targeting webcams using libv4l.
[*] Update the video stabilization plugins to version 0.61.
[*] Added the -X option to tcdecode (let the user specify the acceleration)
[+] Enable versioned, parallel installation.
===========================================================================


import/tcdecode.c
/*
 *  tcdecode.c
 *
 *  Copyright (C) Thomas Oestreich - June 2001
 *
 *  This file is part of transcode, a video stream processing tool
 *
 *  transcode is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  transcode is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with GNU Make; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include "aclib/ac.h"
#include "src/transcode.h"
#include "tccore/tcinfo.h"

#include "libtcutil/xio.h"

#include "ioaux.h"
#include "tc.h"

#include <limits.h>

#define EXE "tcdecode"

extern long fileinfo(int fd, int skip);

int verbose = TC_QUIET;

void import_exit(int code)
{
    if (verbose & TC_DEBUG)
        tc_log_msg(EXE, "(pid=%d) exit (code %d)", (int) getpid(), code);
    exit(code);
}

void version(void)
{
    /* print id string to stderr */
    fprintf(stderr, "%s (%s v%s) (C) 2001-2003 Thomas Oestreich,"
                                    " 2003-2010 Transcode Team\n",
                    EXE, PACKAGE, VERSION);
}

/*************************************************************************/

struct decode_handle {
    const char *name;
    int codec;
    void (*decoder)(decode_t *d);
};

static const struct decode_handle handlers[] = {
    { "mpeg2",   TC_CODEC_MPEG2,   decode_mpeg2 },
    { "ogg",     TC_CODEC_VORBIS,  decode_ogg   },
    { "ac3",     TC_CODEC_AC3,     decode_a52   },
    { "mp3",     TC_CODEC_MP3,     decode_mp3   },
    { "mp2",     TC_CODEC_MP3,     decode_mp2   },
    /* codec intentionally identical */
    { "dv",      TC_CODEC_DV,      decode_dv    },
    { "yuv420p", TC_CODEC_YUV420P, decode_yuv   },
    { "mov",     TC_CODEC_UNKNOWN, decode_mov   },
    /* intentionally unknwon */
    { "lzo",     TC_CODEC_UNKNOWN, decode_lzo   },
    /* intentionally unknwon: lzo1 or lzo2? */
    { "ulaw",    TC_CODEC_ULAW,    decode_ulaw  },
    { NULL,      TC_CODEC_UNKNOWN, NULL },
};

static int decode_stream(const char *codec, decode_t *d)
{
    int i, done = 0;

    for (i = 0; handlers[i].name != NULL; i++) {
        if (!strcmp(codec, handlers[i].name)) {
	        d->codec = handlers[i].codec;
            handlers[i].decoder(d);

            done = 1;
            break;
        }
    }
    return done;
}

/*************************************************************************/

/* ------------------------------------------------------------
 *
 * print a usage/version message
 *
 * ------------------------------------------------------------*/

static void usage(int status)
{
    version();

    fprintf(stderr,"\nUsage: %s [options]\n", EXE);

    fprintf(stderr,"    -i file           input file [stdin]\n");
    fprintf(stderr,"    -x codec          source codec (required)\n");
    fprintf(stderr,"    -t package        codec package\n");
    fprintf(stderr,"    -g wxh            stream frame size [autodetect]\n");
    fprintf(stderr,"    -y format         output raw stream format [rgb]\n");
    fprintf(stderr,"    -Q mode           decoding quality (0=fastest-5=best) [%d]\n", VQUALITY);
    fprintf(stderr,"    -d mode           verbosity mode\n");
    fprintf(stderr,"    -s c,f,r          audio gain for ac3 downmixing [1,1,1]\n");
    fprintf(stderr,"    -A n              A52 decoder flag [0]\n");
    fprintf(stderr,"    -C s,e            decode only from start to end ((V) frames/(A) bytes) [all]\n");
    fprintf(stderr,"    -Y                use libdv YUY2 decoder mode\n");
    fprintf(stderr,"    -z r              convert zero padding to silence\n");
    fprintf(stderr,"    -X type[,type]    override CPU acceleration flags (for debugging)\n");
    fprintf(stderr,"    -v                print version\n");

    exit(status);
}

/* ------------------------------------------------------------
 *
 * universal decode thread frontend
 *
 * ------------------------------------------------------------*/

#define CHECK_OPT do { \
    if (optarg[0] == '-') \
        usage(EXIT_FAILURE); \
} while (0)

int main(int argc, char *argv[])
{
    const char *codec = NULL, *format = "rgb", *magic = "none";
    decode_t decode;
    int ch, done=0;

    memset(&decode, 0, sizeof(decode));
    decode.magic          = TC_MAGIC_UNKNOWN;
    decode.stype          = TC_STYPE_UNKNOWN;
    decode.quality        = VQUALITY;
    decode.ac3_gain[0]    = 1.0;
    decode.ac3_gain[1]    = 1.0;
    decode.ac3_gain[2]    = 1.0;
    decode.frame_limit[0] = 0;
    decode.frame_limit[1] = LONG_MAX;
    decode.accel   = AC_ALL;

    libtc_init(&argc, &argv);

    while ((ch = getopt(argc, argv, "Q:t:d:x:i:a:g:vy:s:YC:A:X:z:?h")) != -1) {
        switch (ch) {
          case 'i':
            CHECK_OPT;
	        decode.name = optarg;
            break;
          case 'd':
            CHECK_OPT;
	        verbose = atoi(optarg);
            break;
          case 'Q':
            CHECK_OPT;
	        decode.quality = atoi(optarg);
            break;
          case 'A':
            CHECK_OPT;
	        decode.a52_mode = atoi(optarg);
            break;
	      case 'x':
            CHECK_OPT;
	        codec = optarg;
            break;
          case 't':
            CHECK_OPT;
	        magic = optarg;
            break;
          case 'y':
            CHECK_OPT;
	        format = optarg;
            break;
          case 'g':
            CHECK_OPT;
	        if (2 != sscanf(optarg,"%dx%d", &decode.width, &decode.height))
                usage(EXIT_FAILURE);
            break;
          case 'v':
            version();
            exit(0);
	        break;
          case 'Y':
            decode.dv_yuy2_mode=1;
            break;
          case 's':
            CHECK_OPT;
	        if (3 != sscanf(optarg,"%lf,%lf,%lf",
                            &decode.ac3_gain[0], &decode.ac3_gain[1], &decode.ac3_gain[2]))
                usage(EXIT_FAILURE);
            break;
          case 'C':
            CHECK_OPT;
	        if (2 != sscanf(optarg,"%ld,%ld",
                            &decode.frame_limit[0], &decode.frame_limit[1]))
                usage(EXIT_FAILURE);
            if (decode.frame_limit[0] >= decode.frame_limit[1]) {
                tc_log_error(EXE, "Invalid -C options");
                usage(EXIT_FAILURE);
	        }
	        break;
          case 'X':
            CHECK_OPT;
            int parsed = ac_parseflags(optarg, &(decode.accel));
            if (!parsed) {
                tc_log_error(EXE, "Invalid -X options");
                usage(EXIT_FAILURE);
            }
          case 'z':
            CHECK_OPT;
	        decode.padrate = atoi(optarg);
            break;
          case 'h':
            usage(EXIT_SUCCESS);
          default:
            usage(EXIT_FAILURE);
        }
    }

    ac_init(decode.accel);

    /* ------------------------------------------------------------
     * fill out defaults for info structure
     * ------------------------------------------------------------*/

    // assume defaults
    if (decode.name == NULL)
        decode.stype = TC_STYPE_STDIN;

    // no autodetection yet
    if (codec == NULL) {
        tc_log_error(EXE, "codec must be specified");
    	usage(EXIT_FAILURE);
    }

    // do not try to mess with the stream
    if (decode.stype == TC_STYPE_STDIN) {
        decode.fd_in = STDIN_FILENO;
    } else {
	    if (tc_file_check(decode.name))
            exit(1);
        decode.fd_in = xio_open(decode.name, O_RDONLY);
        if (decode.fd_in < 0) {
    	    tc_log_perror(EXE, "open file");
	        exit(1);
	    }

	    // try to find out the filetype
    	decode.magic = fileinfo(decode.fd_in, 0);
	    if (verbose)
	        tc_log_msg(EXE, "(pid=%d) %s", getpid(), filetype(decode.magic));
    }

    decode.fd_out  = STDOUT_FILENO;
    decode.codec   = TC_CODEC_UNKNOWN;
    decode.verbose = verbose;
    decode.width   = TC_MAX(0, decode.width);
    decode.height  = TC_MAX(0, decode.height);

    /* ------------------------------------------------------------
     * output raw stream format
     * ------------------------------------------------------------*/

    if (!strcmp(format, "rgb")) decode.format = TC_CODEC_RGB24;
    else if (!strcmp(format, "yuv420p")) decode.format = TC_CODEC_YUV420P;
    else if (!strcmp(format, "yuv2")) decode.format = TC_CODEC_YUV2;
    else if (!strcmp(format, "yuy2")) decode.format = TC_CODEC_YUY2;
    else if (!strcmp(format, "pcm")) decode.format = TC_CODEC_PCM;
    else if (!strcmp(format, "raw")) decode.format = TC_CODEC_RAW;

    /* ------------------------------------------------------------
     * codec specific section
     * note: user provided values overwrite autodetection!
     * ------------------------------------------------------------*/

    // FFMPEG can decode a lot
    if(!strcmp(magic, "ffmpeg") || !strcmp(magic, "lavc")) {
        /* FIXME: swotch to tabel or, better, to tccodecs facilities */
    	if (!strcmp(codec, "mpeg2"))           decode.codec = TC_CODEC_MPEG2;
	    else if (!strcmp(codec, "mpeg2video")) decode.codec = TC_CODEC_MPEG2;
    	else if (!strcmp(codec, "mpeg1video")) decode.codec = TC_CODEC_MPEG1;
    	else if (!strcmp(codec, "divx3"))      decode.codec = TC_CODEC_DIVX3;
    	else if (!strcmp(codec, "divx"))       decode.codec = TC_CODEC_DIVX4;
    	else if (!strcmp(codec, "divx4"))      decode.codec = TC_CODEC_DIVX4;
    	else if (!strcmp(codec, "mp42"))       decode.codec = TC_CODEC_MP42;
    	else if (!strcmp(codec, "mjpg"))       decode.codec = TC_CODEC_MJPEG;
        else if (!strcmp(codec, "mjpeg"))      decode.codec = TC_CODEC_MJPEG;
    	else if (!strcmp(codec, "rv10"))       decode.codec = TC_CODEC_RV10;
    	else if (!strcmp(codec, "svq1"))       decode.codec = TC_CODEC_SVQ1;
    	else if (!strcmp(codec, "svq3"))       decode.codec = TC_CODEC_SVQ3;
    	else if (!strcmp(codec, "vp3"))        decode.codec = TC_CODEC_VP3;
    	else if (!strcmp(codec, "4xm"))        decode.codec = TC_CODEC_4XM;
    	else if (!strcmp(codec, "wmv1"))       decode.codec = TC_CODEC_WMV1;
    	else if (!strcmp(codec, "wmv2"))       decode.codec = TC_CODEC_WMV2;
    	else if (!strcmp(codec, "hfyu"))       decode.codec = TC_CODEC_HUFFYUV;
    	else if (!strcmp(codec, "indeo3"))     decode.codec = TC_CODEC_INDEO3;
    	else if (!strcmp(codec, "h263p"))      decode.codec = TC_CODEC_H263P;
    	else if (!strcmp(codec, "h263i"))      decode.codec = TC_CODEC_H263I;
    	else if (!strcmp(codec, "dvvideo"))    decode.codec = TC_CODEC_DV;
    	else if (!strcmp(codec, "dv"))         decode.codec = TC_CODEC_DV;
    	else if (!strcmp(codec, "vag"))        decode.codec = TC_CODEC_VAG;

	    decode_lavc(&decode);
    }

    done = decode_stream(codec, &decode);

    if (!done) {
	    tc_log_error(EXE, "(pid=%d) unable to handle codec %s", getpid(), codec);
	    exit(1);
    }

    if (decode.fd_in != STDIN_FILENO)
        xio_close(decode.fd_in);

    return 0;
}

#include "libtcutil/static_xio.h"

/*************************************************************************/

/*
 * Local variables:
 *   c-file-style: "stroustrup"
 *   c-file-offsets: ((case-label . *) (statement-case-intro . *))
 *   indent-tabs-mode: nil
 * End:
 *
 * vim: expandtab shiftwidth=4:
 */
